### PR TITLE
[SPARK-9469][SQL] Move Unsafe conversion out of TungstenSort

### PR DESCRIPTION
**Pull request by rxin**
TungstenSort assumes input rows are safe rows, and uses a projection to turn the safe rows into UnsafeRows. This patch takes that part of the logic out of TungstenSort, and move it into the planner. If input rows are already UnsafeRows (e.g. coming out of a TungstenProject), no conversion is needed at all.

### CODE REVIEW

**Comment by rxin** [comment]
cc @JoshRosen

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala [hunk 1]
```diff
@@ -272,11 +272,11 @@ private[joins] final class UnsafeHashedRelation(
     while (i < nKeys) {
       val keySize = in.readInt()
       val valuesSize = in.readInt()
-      if (keySize > keyBuffer.size) {
+      if (keySize > keyBuffer.length) {
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: unrelated change, but didn't want to submit a patch for 2 lines

cc @davies: we should call .length on arrays, instead of size, since .size actually creates a wrapper object, and then calls size on that ...

Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: Thanks

File: sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala [hunk 2]
```diff
@@ -229,7 +229,7 @@ private[spark] object SQLConf {
       " a specific query.")
 
   val UNSAFE_ENABLED = booleanConf("spark.sql.unsafe.enabled",
-    defaultValue = Some(false),
+    defaultValue = Some(true),
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala]: Is this change on purpose?

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala]: ah no
